//
// {C} Copyright 2020 Pensando Systems Inc. All rights reserved
//
//------------------------------------------------------------------------------
///
/// \file
/// protobuf specification for techsupport
///
//------------------------------------------------------------------------------

syntax = "proto3";
package operd;

import "gogo.proto";
import "meta/meta.proto";
import "types.proto";

service TechSupportSvc {
  // TechSupport helps collecting all the debug information on the
  // Distrubuted Services Card (DSC) platform for troubleshooting. The collected
  // information includes trace files from various components, syslogs, core
  // files, output of various debug commands etc.

  // TechSupportCollect initiates a new techsupport information collection
  // NOTE: This API will fail, if there is already one pending request in
  //       progress
  rpc TechSupportCollect (TechSupportRequest) returns (TechSupportResponse) {}
}

// TechSupportSpec defines the configuration of the techsupport object
message TechSupportSpec {
  // SkipCores can be set to true to exclude core files from techsupport
  // output
  bool SkipCores = 1;
}

// TechSupportStatus captures the current status of the techsupport job
message TechSupportStatus {
  // full path of the gzipped techsupport file created, if successful
  string FilePath  = 1;
}

// TechSupportStats contains stats related to techsupport
message TechSupportStats {
}

// Techsupport object
message TechSupport {
  meta.TypeMeta     TypeMeta = 1 [(gogoproto.embed) = true, (gogoproto.jsontag) = ",inline"];
  meta.ObjMeta      ObjMeta  = 2 [(gogoproto.embed) = true, (gogoproto.jsontag) = "meta,omitempty"];
  TechSupportSpec   Spec     = 3 [(gogoproto.jsontag) = "spec,omitempty"];
  TechSupportStatus Status   = 4 [(gogoproto.jsontag) = "status,omitempty"];
  TechSupportStats  Stats    = 5 [(gogoproto.jsontag) = "stats,omitempty"];
}

// TechSupport create request
message TechSupportRequest {
  // config specification
  TechSupportSpec Request  = 1;
}

// TechSupport create response
message TechSupportResponse {
  // success or failure status
  types.ApiStatus   ApiStatus = 1;
  TechSupportStatus Response  = 2;
}
